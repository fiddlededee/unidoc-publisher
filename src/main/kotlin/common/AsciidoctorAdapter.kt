package common

import model.*
import reader.HtmlNode
import reader.UnknownTagProcessing
import writer.*

object AsciidoctorAdapter : GenericAdapter {
    enum class Color { RED, GREEN, YELLOW, BLUE, ORANGE }

    var lightColors = arrayOf(
        Color.RED to "#f8ced0", Color.GREEN to "#c2f3d3",
        Color.YELLOW to "#fef0c2", Color.BLUE to "#c2f3f1",
        Color.ORANGE to "#fedaca"
    )
    var darkColors = arrayOf(
        Color.RED to "#8f141b", Color.GREEN to "#19803d",
        Color.YELLOW to "#a17c02", Color.BLUE to "#19807c",
        Color.ORANGE to "#a03203"
    )
    var admonitionColors = arrayOf(
        "note" to Color.BLUE,
        "tip" to Color.GREEN,
        "important" to Color.YELLOW,
        "caution" to Color.ORANGE,
        "warning" to Color.RED
    )

    enum class AdmonitionCellType { TYPE_CONTENT, CONTENT }

    override val unknownTagProcessingRule: HtmlNode.() -> UnknownTagProcessing = {
        if (classNames().intersect(
                setOf(
                    "listingblock",
                    "admonitionblock",
                    "exampleblock",
                    "imageblock",
                    "paragraph",
                    "book",
                    "article",
                    "title",
                    "footnotes",
                    "footnote",
                    "dlist",
                    "olist",
                    "ulist",
                    "content",
                    *(1..6).map { "sect$it" }.toTypedArray(),
                )
            ).isNotEmpty()
        ) {
            UnknownTagProcessing.PROCESS
        } else if (setOf("dl", "dt", "dd").contains(nodeName())) {
            UnknownTagProcessing.PROCESS
        } else if (nodeName() == "div") {
            UnknownTagProcessing.PASS
        } else UnknownTagProcessing.UNDEFINDED
    }

    override fun basicStyle(): OdtStyleList {
        return OdtStyleList(
            OdtStyle { genericTitle ->
                if (genericTitle !is Paragraph) return@OdtStyle
                val genericTitleWrapper = genericTitle.parent()
                if (genericTitleWrapper == null ||
                    !genericTitleWrapper.roles.contains("title")
                ) return@OdtStyle
                attributes("text:style-name" to "Text")
            },
            OdtStyle { tableCell ->
                if (tableCell !is TableCell) return@OdtStyle
                tableCellProperties {
                    arrayOf("top", "right", "bottom", "left").forEach {
                        attribute("fo:border-$it", "0.5pt solid #000000")
                        if (it != "bottom") attribute("fo:padding-$it", "1mm")
                    }
                    if (tableCell.sourceTagName == "th") attribute("fo:background-color", "#eeeeee")
                }
                arrayOf("top", "middle", "bottom").forEach {
                    if (tableCell.roles.contains("valign-$it"))
                        tableCellProperties { attributes("style:vertical-align" to it) }
                }
            },
            OdtStyle { paragraph ->
                if (paragraph !is Paragraph) return@OdtStyle
                if (paragraph.ancestor { it is TableCell }.isEmpty()) return@OdtStyle
                attributes("text:style-name" to "Table Contents")
            },
            OdtStyle { paragraph ->
                if (paragraph !is Paragraph) return@OdtStyle
                arrayOf("center", "left", "right").forEach {
                    if (paragraph.parent()?.roles?.contains("halign-$it") == true)
                        paragraphProperties { attributes("fo:text-align" to it) }
                }
            },
            OdtStyle { tableCell ->
                if (tableCell !is TableCell) return@OdtStyle
                val nodeParent = tableCell.parent() ?: return@OdtStyle
                lightColors.forEach { color ->
                    if (nodeParent.descendant()
                            .map { it.roles }.reduce { acc, roles ->
                                ArrayList(acc + roles)
                            }.contains("row-background-color-${color.first.toString().lowercase()}")
                    ) {
                        tableCellProperties { attributes("fo:background-color" to color.second) }
                    }
                }

            },
            OdtStyle { table ->
                if (table !is Table) return@OdtStyle
                if (table.ancestor { it is Table }.isEmpty())
                    tableProperties { attributes("fo:margin-bottom" to "2mm") } else
                    tableProperties { attributes("fo:margin-bottom" to "1mm") }
            },
            OdtStyle { admonitionTypeContentCell ->
                if (admonitionTypeContentCell !is TableCell) return@OdtStyle
                val admonitionCellType = run {
                    if (admonitionTypeContentCell.roles.contains("admonition-type-content")) {
                        AdmonitionCellType.TYPE_CONTENT
                    } else if (admonitionTypeContentCell.roles.contains("admonition-content"))
                        AdmonitionCellType.CONTENT
                    else null
                }
                if (admonitionCellType == null) return@OdtStyle
                val admonitionBlockRoles = admonitionTypeContentCell
                    .ancestor { it.roles.contains("admonitionblock") }
                    .firstOrNull()
                    ?.roles ?: throw Exception("Not in an admonition block")
                admonitionColors.filter { admonitionBlockRoles.contains(it.first) }
                    .first()
                    .apply {
                        tableCellProperties {
                            val color =
                                lightColors.toMap()[second]
                                    ?: throw Exception("Color ${second} not found")
                            if (admonitionCellType == AdmonitionCellType.TYPE_CONTENT)
                                attributes("fo:background-color" to color)
                            arrayOf("left", "right", "top", "bottom").forEach {
                                attributes("fo:border-$it" to "0.7pt solid $color")
                            }
                        }
                    }
            },
            OdtStyle { admonitionTitleParagraph ->
                if (admonitionTitleParagraph !is Paragraph) return@OdtStyle
                if (admonitionTitleParagraph.ancestor { it.roles.contains("admonition-content") }.isEmpty())
                    return@OdtStyle
                if (admonitionTitleParagraph.ancestor { it.roles.contains("title") }.isEmpty())
                    return@OdtStyle
                textProperties { attributes("fo:font-style" to "italic") }
            },
            OdtStyle { titleParagraph ->
                if (titleParagraph !is Paragraph) return@OdtStyle
                if (!titleParagraph.roles.contains("doctitle")) return@OdtStyle
                attributes("text:style-name" to "Title")
            },
            OdtStyle { figureTitle ->
                if (figureTitle !is Paragraph) return@OdtStyle
                if (figureTitle.ancestor { it.roles.contains("imageblock") }.isEmpty()) return@OdtStyle
                if (figureTitle.ancestor { it.roles.contains("title") }.isEmpty()) return@OdtStyle
                attributes("text:style-name" to "Figure")
            },
            OdtStyle { imageblockParagraph ->
                if (imageblockParagraph !is Paragraph) return@OdtStyle
                if (!imageblockParagraph.roles.contains("imageblock-paragraph"))
                    return@OdtStyle
                attributes("text:style-name" to "Image Block")
                val imageBlockRoot =
                    imageblockParagraph.ancestor { it.roles.contains("imageblock") }
                        .firstOrNull()
                if (imageBlockRoot != null &&
                    imageBlockRoot.descendant { it.roles.contains("title") }.isNotEmpty()
                ) paragraphProperties { attributes("fo:keep-with-next" to "always") }
            },
            OdtStyle { tableTitle ->
                if (tableTitle !is Paragraph) return@OdtStyle
                val parent = tableTitle.parent()
                if (parent != null && parent.roles.contains("table-title")) {
                    attributes("text:style-name" to "Table")
                }
            },
            OdtStyle { listTitle ->
                if (listTitle !is Paragraph) return@OdtStyle
                val listTitleWrapper = listTitle.parent() ?: return@OdtStyle
                if (!listTitleWrapper.roles.contains("title")) return@OdtStyle
                if (!listTitleWrapper.hasNext()) return@OdtStyle
                val listTitleWrapperNextSibling = listTitleWrapper.next()
                if (listTitleWrapperNextSibling !is UnorderedList &&
                    listTitleWrapperNextSibling !is OrderedList
                ) return@OdtStyle
                if (listTitle.ancestor { it is Table }.isEmpty())
                    attributes("text:style-name" to "Text body") else
                    attributes("text:style-name" to "Table Contents")
                paragraphProperties { attributes("fo:keep-with-next" to "always") }
            },
            OdtStyle { listingTitle ->
                if (listingTitle !is Paragraph) return@OdtStyle
                val listingTitleWrapper = listingTitle.parent() ?: return@OdtStyle
                if (!listingTitleWrapper.roles.contains("title")) return@OdtStyle
                if (!listingTitleWrapper.hasNext()) return@OdtStyle
                val listingTitleWrapperNextSibling = listingTitleWrapper.next()
                if (!listingTitleWrapperNextSibling.roles.contains("pre")) return@OdtStyle
                attributes("text:style-name" to "Text")
            },
            OdtStyle { paragraph ->
                if (paragraph !is Paragraph) return@OdtStyle
                if (!paragraph.roles.contains("definition-term")) return@OdtStyle
                attributes("text:style-name" to "Definition Term")
            },
            OdtStyle { paragraph ->
                if (paragraph !is Paragraph) return@OdtStyle
                if (!paragraph.roles.contains("horizontal-line-for-example")) return@OdtStyle
                attributes("text:style-name" to "Horizontal Line for Example")
            },
            OdtStyle { node ->
                if (node.ancestor { it.roles.contains("keep-with-next") }.isNotEmpty()
                    || node.roles.contains("keep-with-next")
                ) {
                    if (node is Paragraph)
                        paragraphProperties { attributes("fo:keep-with-next" to "always") }
                    if (node is Table)
                        tableProperties { attributes("fo:keep-with-next" to "always") }
                    if (node is TextFrame)
                        graphicProperties { attributes("fo:keep-with-next" to "always") }
                }
            },
            OdtStyle { span ->
                if (span !is Span) return@OdtStyle
                darkColors.forEach {
                    if (span.roles.contains(it.first.toString().lowercase())) {
                        textProperties { attributes("fo:color" to it.second) }
                    }
                }
            },
            OdtStyle { paragraph ->
                if (paragraph !is Paragraph) return@OdtStyle
                if (!paragraph.roles.contains("text-frame-wrapper-paragraph")) return@OdtStyle
                textProperties { attributes("fo:font-size" to "0pt") }
                paragraphProperties { attributes("fo:font-size" to "0pt") }
            }
        )
    }

    fun Node.trimSimpleTdContent() {
        this.descendant { it is TableCell }.map { it as TableCell }
            .forEach { tableCell ->
                if (tableCell.children().size == 1 && tableCell.children()[0] is Text) {
                    val textNode = tableCell.children()[0] as Text
                    textNode.text = textNode.text.trim()
                }
            }
    }

    @JvmStatic
    fun Node.wrapTableCellInlineContent() {
        this.descendant { it is TableCell }.map { it as TableCell }
            .forEach { tableCell -> tableCell.wrapNodeInlineContents() }
    }

    @JvmStatic
    fun Node.wrapTitleInlineContent() {
        this.descendant { it is OpenBlock && it.roles.contains("title") }
            .forEach { openBlock ->
                openBlock.wrapNodeInlineContents()
                val table = openBlock.parent()
                if (table != null && table is Table) {
                    val tableId = table.id
                    if (tableId != null) {
                        openBlock.descendant { it is Paragraph }.firstOrNull()?.apply {
                            id = tableId
                        }
                        table.id = "table--${tableId}"
                    }
                    table.insertBefore(openBlock)
                    openBlock.roles("table-title")
                }
            }
    }

    @JvmStatic
    fun Node.extractToc() {
        val parsedToc = this.descendant { it.id == "toc" }
            .firstOrNull() ?: return
        val toc = Toc().apply {
            levels = parsedToc.descendant { it is UnorderedList }
                .maxOfOrNull { ul -> ul.ancestor { it is UnorderedList }.size + 1 }
                ?: throw Exception("Can't determine number of levels in the TOC")
            titleNode = Paragraph().apply {
                parsedToc.descendant { it.parent()?.id == "toctitle" }.forEach {
                    appendChild(it)
                }
            }
        }
        parsedToc.insertBefore(toc)
        parsedToc.remove()
    }


    fun Node.transformAdmonitions() {
        this.descendant { it.roles.contains("admonitionblock") }.forEach { block ->
            val oldTable = block.children().first { it is Table }
            val (typeContent, content) = run {
                oldTable
                    .children().first { it is TableRowGroup }
                    .children().first { it is TableRow }
                    .apply {
                        if (children().size != 2 ||
                            children().any { it !is TableCell }
                        ) throw Exception("Can't define admonition block")
                    }.children()
            }
            typeContent.roles("admonition-type-content")
            content.roles("admonition-content")
            block.apply {
                table {
                    col(Length(100F))
                    tableRowGroup(TRG.body) {
                        tr { appendChild(typeContent) }
                        tr { appendChild(content) }
                    }
                }
            }
            oldTable.remove()
        }
    }

    @JvmStatic
    fun Node.normalizeHeaders() {
        this.descendant { it is Header }.map { it as Header }.forEach { header ->
            if (header.level == 1) {
                val titleParagraph = Paragraph().apply {
                    roles("doctitle")
                    header.children().forEach { appendChild(it) }
                }
                header.insertBefore(titleParagraph)
                header.remove()
            } else header.level -= 1
        }
    }

    @JvmStatic
    fun Node.wrapTablesInLists() {
        this.descendant { it is ListItem }.forEach { listItem ->
            listItem.descendant().forEach liChildren@{ table ->
                if (table !is Table) return@liChildren
                if (table.parent() is TextFrame) return@liChildren
                val textFrame = TextFrame()
                val textFrameWrapperParagraph = Paragraph().apply {
                    roles("text-frame-wrapper-paragraph")
                    appendChild(textFrame)
                }
                table.insertBefore(textFrameWrapperParagraph)
                textFrame.appendChild(table)
            }
        }
    }


    @JvmStatic
    fun Node.extractFootnotes() {
        val footnotes: MutableMap<String, Footnote> = mutableMapOf()
        this.descendant { it is Span && it.roles.contains("footnote") }
            .forEach { footNoteSpan ->
                val footnoteId = footNoteSpan.id ?: throw Exception("Footnote id not found")
                val footnoteHref = footNoteSpan.descendant { it is Anchor }.map { it as Anchor }
                    .firstOrNull()
                    .apply { if (this == null) throw Exception("Anchor to footnote not found") }!!
                    .href
                    .substring(1)
                val footnoteBody = footNoteSpan.ancestor().last()
                    .descendant { it.id == footnoteHref }
                    .firstOrNull()
                    .apply { if (this == null) throw Exception("Footnote $footnoteHref not found") }!!
                    .children()
                    .mapIndexed { i, el ->
                        if (i == 0 && el is Anchor) {
                            null
                        } else if (i == 1 && el is Text) {
                            Text(el.text.substring(2))
                        } else {
                            el
                        }
                    }.filterNotNull()
                val footnote = Footnote(footnoteId).apply {
                    footnoteBody.forEach { this.appendChild(it) }
                }
                footNoteSpan.insertAfter(footnote)
                footNoteSpan.remove()
                footnotes[footnoteHref] = footnote
            }
        this.descendant { it is Span && it.roles.contains("footnoteref") }
            .forEach { footnoteRefSpan ->
                val footnoteToRefHref = footnoteRefSpan.descendant { it is Anchor }.map { it as Anchor }
                    .firstOrNull()
                    .apply { if (this == null) throw Exception("Anchor to footnote reference not found") }!!
                    .href
                    .substring(1)
                val footnoteToRef = footnotes[footnoteToRefHref]
                    ?: throw Exception(
                        "Can't find footnote ${footnoteRefSpan.id} to reference"
                    )
                val footnoteRef = FootnoteRef(footnoteToRef)
                footnoteRefSpan.insertAfter(footnoteRef)
                footnoteRefSpan.remove()
            }
    }

    fun Node.wrapBlockImages() {
        this.descendant { image ->
            image is Image &&
                    image.ancestor { it.roles.contains("imageblock") }
                        .isNotEmpty()
        }.forEach { image ->
            val imageBlock = image.ancestor { it.roles.contains("imageblock") }.firstOrNull()
            val p =
                Paragraph().apply { roles("imageblock-paragraph") }
            if (imageBlock?.id != null) {
                p.id = imageBlock.id
                imageBlock.id = "image-block--${imageBlock.id}"
            }
            image.insertAfter(p)
            p.appendChild(image)
        }
    }

    fun Node.transformDefinitionList() {
        this.descendant { it.sourceTagName == "dt" }.forEach { dt ->
            dt.apply {
                val paragraph = Paragraph().apply {
                    dt.children().forEach { appendChild(it) }
                    roles("definition-term")
                }
                insertAfter(paragraph)
                remove()
            }
        }
    }

    fun Node.transformExamples() {
        descendant { it.roles.contains("exampleblock") }.forEach { exampleBlock ->
            exampleBlock.descendant { it.roles.contains("content") }.forEach { exampleContent ->
                exampleContent.appendFirstChild(
                    Paragraph().apply {
                        roles("keep-with-next", "horizontal-line-for-example")
                    }
                )
                exampleContent.p { roles("horizontal-line-for-example") }
                    .previous().roles("keep-with-next")
            }
        }
    }

    override fun Node.normalizeAll() {
        trimSimpleTdContent()
        extractToc()
        wrapTableCellInlineContent()
        wrapTitleInlineContent()
        wrapBlockImages()
        normalizeHeaders()
        transformAdmonitions()
        transformExamples()
        transformDefinitionList()
        extractFootnotes()
        wrapTablesInLists()
        normalizeImageDimensions()
        setImageBestFitDimensions()
    }
}