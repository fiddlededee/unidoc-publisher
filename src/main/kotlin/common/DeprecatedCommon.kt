package common

@Deprecated("Renamed to converter.fodt.AsciidoctorOdAdapter")
typealias AsciidoctorAdapter = converter.fodt.AsciidoctorOdAdapter

