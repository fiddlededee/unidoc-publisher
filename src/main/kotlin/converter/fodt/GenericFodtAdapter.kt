package converter.fodt

import model.Node
import reader.HtmlNode
import reader.UnknownTagProcessing
import writer.OdtStyleList

interface GenericFodtAdapter {
    val unknownTagProcessingRule: HtmlNode.() -> UnknownTagProcessing
    fun basicStyle(): OdtStyleList
    fun Node.normalizeAll() {}
}