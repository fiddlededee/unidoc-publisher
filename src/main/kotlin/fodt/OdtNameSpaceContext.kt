package fodt

import javax.xml.namespace.NamespaceContext

class OdtNameSpaceContext : NamespaceContext {
    override fun getNamespaceURI(p0: String?): String? {
        val namespaceURI = when (p0) {
            "text" -> "urn:oasis:names:tc:opendocument:xmlns:text:1.0"
            "style" -> "urn:oasis:names:tc:opendocument:xmlns:style:1.0"
            "fo" -> "urn:oasis:names:tc:opendocument:xmlns:xsl-fo-compatible:1.0"
            "office" -> "urn:oasis:names:tc:opendocument:xmlns:office:1.0"
            "table" -> "urn:oasis:names:tc:opendocument:xmlns:table:1.0"
            "draw" -> "urn:oasis:names:tc:opendocument:xmlns:drawing:1.0"
            "svg" -> "http://www.w3.org/2000/svg"
            "loext" -> "urn:org:documentfoundation:names:experimental:office:xmlns:loext:1.0"
            else -> null
        }
        return namespaceURI
    }

    override fun getPrefix(p0: String?): String {
        throw UnsupportedOperationException()
    }

    override fun getPrefixes(p0: String?): MutableIterator<String> {
        throw UnsupportedOperationException()
    }
}

