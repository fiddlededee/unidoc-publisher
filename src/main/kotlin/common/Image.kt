package common

import fodt.iterable
import fodt.parseStringAsXML
import fodt.xpath
import model.LengthUnit
import org.w3c.dom.Attr
import java.util.*
import javax.imageio.ImageIO
import java.io.ByteArrayInputStream
import java.io.File
import javax.imageio.IIOException

data class ImageMeta(val width: Float, val height: Float, val unit: LengthUnit)

fun imageMeta(src: String): ImageMeta {
    val base64Regex = """^data:image/(.*);base64,(.*)$""".toRegex()
    val matchResult = base64Regex.matchEntire(src)
    val type = if (matchResult == null) src.substringAfterLast('.', "") else
        matchResult.groupValues[1]
    return if (type == "svg" || type == "svg+xml") svgImageMeta(matchResult, src) else
        binaryImageMeta(matchResult, src)
}

private fun svgImageMeta(matchResult: MatchResult?, src: String): ImageMeta {
    val svgXml = run {
        if (matchResult == null) {
            File(src).readText()
        } else {
            val base64Image = matchResult.groupValues[2]
            String(Base64.getDecoder().decode(base64Image))
        }
    }
    val viewBox = svgXml.parseStringAsXML().xpath("/svg:svg/@viewBox").iterable().map { it as Attr }
        .firstOrNull()?.value
        ?.split(" ") ?: throw Exception("ViewBox not found in svg file")
    return ImageMeta(viewBox[2].toFloat(), viewBox[3].toFloat(), LengthUnit.parrots)
}

private fun binaryImageMeta(matchResult: MatchResult?, src: String): ImageMeta {
    val img = run {
        if (matchResult == null) {
            try {
                ImageIO.read(File(src))
            }catch (e : IIOException) {
                throw IIOException("Can't read input file ${src}")
            }
        } else {
            val base64Image = matchResult.groupValues[2]
            val imageBytes: ByteArray = Base64.getDecoder().decode(base64Image)
            ImageIO.read(ByteArrayInputStream(imageBytes))
        }
    }
    return ImageMeta(img.width.toFloat(), img.height.toFloat(), LengthUnit.px)
}
