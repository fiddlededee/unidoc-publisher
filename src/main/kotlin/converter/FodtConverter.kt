package converter

import common.GenericAdapter
import fodt.FodtGenerator
import model.Document
import model.Node
import model.Text
import org.redundent.kotlin.xml.PrintOptions
import reader.GenericHtmlReader
import reader.HtmlNode
import reader.UnknownTagProcessing
import writer.OdWriter
import writer.OdtStyleList

open class FodtConverter(init: FodtConverter.() -> Unit) {
    private var adapter: GenericAdapter? = null
    fun adaptWith(adapter: GenericAdapter) {
        unknownTagProcessingRule = adapter.unknownTagProcessingRule
        odtStyleList = adapter.basicStyle()
        this.adapter = adapter
    }

    var html: String? = null
        set(value) {
            val fieldNotNullable = value ?: throw Exception("html value can't be null")
            field = fieldNotNullable; htmlNode = HtmlNode(fieldNotNullable)
        }

    var xpath: String = "/html/body"
    var unknownTagProcessingRule: HtmlNode.() -> UnknownTagProcessing = { UnknownTagProcessing.UNDEFINDED }
    var customNodeProcessingRule: Node.(htmlNode: HtmlNode) -> Unit = {}
    var odtStyleList = OdtStyleList()
    var template: String? = null
    var ast: Node? = null
    var preList: ArrayList<PreRegion> = arrayListOf()
    var fodtGenerator: FodtGenerator? = null
    var fodt: String? = null
    private var htmlNode: HtmlNode? = null

    init {
        this.apply(init)
    }

    fun ast(): Node {
        return ast ?: throw Exception("AST is empty")
    }

    fun template(): String {
        return template ?: throw Exception("Template is empty")
    }

    fun preList(): ArrayList<PreRegion> {
        if (preList.isEmpty()) throw Exception("The list of pre FODT elements is empty")
        return preList
    }

    fun html(): String {
        return html ?: throw Exception("HTML is emplty")
    }

    fun fodt(): String {
        return fodt ?: throw Exception("Fodt is empty")
    }

    fun parse() {
        val localHtmlNode = htmlNode?.selectAtXpath(xpath)
            ?: throw Exception("Please set html")
        ast = Document()
        val localAst = ast ?: throw Exception("Error: ast variable was mutated")
        newReaderInstance(
            localAst, localHtmlNode,
            unknownTagProcessingRule, customNodeProcessingRule
        )
            .apply { localAst.setBasics(localHtmlNode); iterateAll() }
        // Don't understand html rendering rules, hope I'm right about this
        localAst.descendant { it is Text && it.text.isBlank() }.forEach { blankTextNode ->
            if (!blankTextNode.hasPrevious() ||
                !blankTextNode.previous().isInline ||
                blankTextNode.previous().roles.contains("br") ||
                !blankTextNode.hasNext() ||
                !blankTextNode.next().isInline ||
                blankTextNode.next().roles.contains("br")
            )
                blankTextNode.remove()
        }
    }

    fun generatePre() {
        val localAst = ast ?: throw Exception("Ast is null")
        adapter?.apply { localAst.normalizeAll() }
        (localAst.descendant { it.includeTags.isNotEmpty() } +
                localAst
                ).forEach { nodeToProcess ->
                val includeTags = nodeToProcess.includeTags
                    .let { if (it.isNotEmpty()) it.toSet() else setOf("all") }
                val pre = newOdWriterInstance(odtStyleList = odtStyleList)
                    .apply { (nodeToProcess.write(this)) }
                    .preOdNode.toString(PrintOptions(pretty = false))
                preList.add(PreRegion(includeTags, pre, nodeToProcess.isInline))
            }

    }

    fun generateFodt() {
        val localTemplate = template ?: throw Exception("Please, set template variable")
        if (preList.isEmpty()) throw Exception("Please set pre variable")
        fodtGenerator = FodtGenerator(preList, localTemplate)
        serializeFodt()
    }

    fun convert() {
        parse(); generatePre(); generateFodt()
    }

    fun ast2fodt() {
        generatePre(); generateFodt()
    }

    fun serializeFodt() {
        fodt = fodtGenerator?.serialize()
    }

    open fun newReaderInstance(
        ast: Node, htmlNode: HtmlNode,
        unknownTagProcessingRule: HtmlNode.() -> UnknownTagProcessing,
        customNodeProcessing: Node.(htmlNode: HtmlNode) -> Unit
    ): GenericHtmlReader {
        return GenericHtmlReader(ast, htmlNode, unknownTagProcessingRule, customNodeProcessing)
    }

    open fun newOdWriterInstance(odtStyleList: OdtStyleList): OdWriter {
        return OdWriter(odtStyleList = odtStyleList)
    }
}