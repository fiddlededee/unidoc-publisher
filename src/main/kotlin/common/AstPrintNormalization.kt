package common

import model.Image
import model.Length
import model.LengthUnit
import model.Node

fun getBestDimensions(
    widthPx: Float,
    heightPx: Float,
    fitRectWidthMmm: Float? = null,
    fitRectHeightMmm: Float? = null,
    srcDpi: Float? = null,
): ImageMeta? {
    if (srcDpi != null) {
        val (dpiExpectedWidth, dpiExpectedHeight) = arrayOf(widthPx, heightPx).map { it / srcDpi * 2540 }
        if ((fitRectWidthMmm == null || fitRectHeightMmm == null) || (dpiExpectedWidth <= fitRectWidthMmm && dpiExpectedHeight <= fitRectHeightMmm)) return ImageMeta(
            dpiExpectedWidth,
            dpiExpectedHeight,
            LengthUnit.cmm
        )
    }
    if (fitRectWidthMmm != null && fitRectHeightMmm != null) {
        if (widthPx / heightPx > fitRectWidthMmm / fitRectHeightMmm) {
            return (ImageMeta(
                fitRectWidthMmm, fitRectWidthMmm / widthPx * heightPx, LengthUnit.cmm
            ))
        } else {
            return (ImageMeta(
                fitRectHeightMmm / heightPx * widthPx, fitRectHeightMmm, LengthUnit.cmm
            ))
        }
    }
    return null
}

fun Node.setImageBestFitDimensions() {
    this.descendant { it is Image }.map { it as Image }.forEach { image ->
        val reFitRect = """^fitrect-([0-9]+)-([0-9]+)$""".toRegex()
        val reSrcDpi = """^srcdpi-([0-9]+)$""".toRegex()
        val thisOrAncestorRoles = arrayListOf<String>()
        image.ancestor().forEach { thisOrAncestorRoles.addAll(it.roles) }
        thisOrAncestorRoles.addAll(image.roles)

        val srcDpi =
            thisOrAncestorRoles.firstNotNullOfOrNull { reSrcDpi.matchEntire(it) }
                ?.run { groupValues[1] }?.toFloat()
        val (fitRectWidthMmm, fitRectHeightMmm) = run {
            val matchResult = thisOrAncestorRoles.firstNotNullOfOrNull { reFitRect.matchEntire(it) }
            if (matchResult != null) {
                arrayOf(
                    matchResult.groupValues[1].toFloat(),
                    matchResult.groupValues[2].toFloat()
                )
            } else arrayOf<Float?>(null, null)
        }
        val realImageMeta = imageMeta(image.src)
        val bestFitImageMeta = getBestDimensions(
            realImageMeta.width, realImageMeta.height,
            fitRectWidthMmm, fitRectHeightMmm, srcDpi
        )
        if (bestFitImageMeta != null) {
            image.width = Length(bestFitImageMeta.width, bestFitImageMeta.unit)
            image.height = Length(bestFitImageMeta.height, bestFitImageMeta.unit)
        }
    }
}

fun Node.normalizeImageDimensions(portraitWidth: Float = 17000F, basicDpi: Float = 96F) {
    this.descendant { it is Image }.map { it as Image }.forEach { image ->
        val (imageWidth, imageHeight) = arrayOf(image.width, image.height)
        arrayOf(imageWidth, imageHeight).filterNotNull().forEach {
            when (it.unit) {
                LengthUnit.perc -> {
                    it.unit = LengthUnit.cmm
                    it.value *= portraitWidth / 100
                }

                LengthUnit.px -> {
                    it.unit = LengthUnit.cmm
                    it.value = it.value * 2540 / basicDpi
                }

                else -> {}
            }
        }
        if (imageHeight == null) {
            val imageMeta = imageMeta(image.src)
            val newImageWidth = run {
                imageWidth ?: Length(imageMeta.width / basicDpi * 2540F, LengthUnit.cmm)
            }
            image.width = newImageWidth
            val proportionalHeight = newImageWidth.value / imageMeta.width * imageMeta.height
            image.height = Length(proportionalHeight, newImageWidth.unit)
        }
    }
}