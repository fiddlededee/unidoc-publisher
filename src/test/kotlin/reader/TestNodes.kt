package reader

import converter.fodt.AsciidoctorOdAdapter
import common.asciidocAsHtml
import common.prettySerialize
import converter.fodt.FodtConverter
import fodt.parseStringAsXML
import model.Document
import org.junit.jupiter.api.Test
import verify


class TestNodes {

    @Test
    fun table() {
        FodtConverter {
            html = "<table><tr><td>A1</td><td>A2</td></tr></table>"
            parse()
            ast?.toYamlString()!!.verify()
        }
    }

    @Test
    fun image() {
        FodtConverter {
            html = "<img src='my-images/image.png' width='10pt'></img>"
            parse()
            ast?.toYamlString()!!.verify()
        }
    }

    @Test
    fun includeTags() {
        FodtConverter {
            html = "<p class='tag--content-1 tag--c2'><span class='tag--c3'>some text</span></p>"
            parse()
            ast?.toYamlString()!!.verify()
        }
    }

    @Test
    fun includeTagsAsciidoc() {
        FodtConverter {
            html = """
                [.tag--content-1]
                Some paragraph
            """.trimIndent().asciidocAsHtml()
            xpath = "/html/body"
            unknownTagProcessingRule = AsciidoctorOdAdapter.unknownTagProcessingRule
            println(html)
            parse()
            ast?.toYamlString()!!
                .replace(""".*Last update.*""".toRegex(), "")
                .verify()
        }

    }

    @Test
    fun IncludeTagsPre() {
        FodtConverter {
            ast = Document().apply {
                p { +"some text 1" }
                p {
                    includeTags = mutableSetOf("content-1")
                    +"some text 2"
                    span {
                        includeTags = mutableSetOf("content-2, content-3")
                    }
                }
            }
            generatePre()
            preList.joinToString("\n") {
                arrayOf(
                    it.includeTags.joinToString(","),
                    it.pre.parseStringAsXML().prettySerialize()
                ).joinToString("\n")
            }.verify()
        }
    }

}