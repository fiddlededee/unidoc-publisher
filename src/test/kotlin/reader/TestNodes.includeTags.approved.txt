!<Document>
sourceTagName: "body"
children:
- !<Paragraph>
  sourceTagName: "p"
  includeTags:
  - "content-1"
  - "c2"
  roles:
  - "tag--content-1"
  - "tag--c2"
  children:
  - !<Span>
    sourceTagName: "span"
    includeTags:
    - "c3"
    roles:
    - "span"
    - "tag--c3"
    children:
    - !<Text>
      text: "some text"
